### PR TITLE
JobService initial design document

### PR DESCRIPTION
**Pull request by jayofdoom**
Still needs questions within answered.

Co-Authored-By: Kevin Hannon <kannon1992@gmail.com>

### CODE REVIEW

**Comment by rankindaveyg** [comment]
Can we rename this something like "job status service"? Caching usually means something different than what is proposed here.

**Comment by d80tb7** [comment]
This looks good- a few comments:

* I think I'd keep the scope of this limited to "a component that makes airflow integration scale" for now. It may be that this becomes more generally useful in the future, but for now I think keeping the scope limited will be helpful.
* Are there some cases where we need more than just a simple job status.  I'm thinking specifically about job failures or rejections where we may need to tell airflow *why* the job has failed.
* I think my preference would be to run this outside of the Armada deployment and would probably favour deployment alongside airflow- i.e. one component per airflow deployment.  I suspect Armada will generally manage more resources than a single Airflow deployment will be able to deal with, which implies that large Armada deployments can support multiple Airflow instances.  By adding this cache at the airflow level, it means we don't have to create a single, giant cache at on the Armada side but end up with a somewhat naturally sharded approach.
* In terms for naming- I guess this sin't really a cache, but rather an "(Airflow) job status service" or similar.

File: docs/cache.md [hunk 1]
```diff
@@ -0,0 +1,82 @@
+# Armada local caching
+
+## Problem Description
+Armada’s API is event driven, preventing it from integrating with tools, such as Apache Airflow, written with the expectation that it can easily fetch status of a running job. It is not scalable to have Airflow subscribe to the event stream to observe status, so we must implement a caching layer which will expose a friendlier API for individual job querying.
+
+## Proposed Change
+### Notes
+- Add an optional caching API and service to Armada
+- Caches job_id:job_status relationship for subscribed (queue,job_set) tuples
+  - Is there other information this cache might need?
```
[the remark this answers is not in the document]
Comment (kannon92) [reply docs/cache.md]: I don't think I'd want that as the API but I think this is a good idea for our data store.

**Comment by michelsumbul** [comment]
Hey guys, thanks for this design doc this is great! 
Few thoughs that we should discuss:

-	The job service / caching should be an optional deployment of armada and not part of airflow. If it’s part of the airflow it will be more difficult to get it accepted by the airflow community. The airflow operator should be a simple client, like any client for any other system. I don’t think we want to put any complexity in the client. What do you think? 

-	What do you mean concretely by one cache per user/aiflow? Is it one cache structure in memory per client in a single server or multiple cache server instance?
Having one per airflow mean we could easily have ten’s of it. It also mean that each time we deploy an airflow we will need to deploy a new cache server which should be 2 thing completely independent.

-	I know this cache server is currently needed for airflow but on a long term we could easily imagine having the need for it for other usercase.

Why not having a distributed cache service for everything? We could imagine multiple server/container with one leader to get info from armada and updating all the follower with a Raft protocol. Read could then be scale out by just adding more container.
it could also be any other distributed cache / ephemeral storage.

-	Authorization: it’s good to think about security, however the cache server if I understand correctly will be read only and should not have confidential info, just job name and status. So we could imagine have a v1 one without security and v2 with security. The security could be implemented by adding to the cache a field with the owner of the job that can fetch job status

-	Airflow also fetch log of the different task in real time (to be verified). How do you plan to address this with the cache server / job status service?

-	Where the cache server fit in the process for airflow to cancel / retry tasks? (worst use case will be to have airflow to cancel huge dag with hundred/thousands of tasks started)

-	Devil advocate: why going the route of a separate service? Could we not improve the current service by supporting also pulling and potentially implementing a distributed cache there directly?

-	It will be good to have few word on the reason why armada will not scale / support high pull load? What are the technical reasons that motivate us of doing this.

For info, we have an internal process to capture design decision call ADR at GR. Once we have this doc in good shape and some consensus, we should put it in the ADR and get formal approval from armada/architect team.

Michel

**Comment by kannon92** [comment]
> Hey guys, thanks for this design doc this is great! Few thoughs that we should discuss:
> 
> * The job service / caching should be an optional deployment of armada and not part of airflow. If it’s part of the airflow it will be more difficult to get it accepted by the airflow community. The airflow operator should be a simple client, like any client for any other system. I don’t think we want to put any complexity in the client. What do you think?

This is probably a good question for @severinson or @d80tb7.  I do think putting the cache in Armada makes the most sense so we aren't reinventing the wheel when we integrate with our schedulers/third-party tools.  I would say that I structured our operator to be flexible with either design because we are just using the JobService as a client.  

Deployment is other issue however.  
> * What do you mean concretely by one cache per user/aiflow? Is it one cache structure in memory per client in a single server or multiple cache server instance?

@jayofdoom What did you mean by this?  
>   Having one per airflow mean we could easily have ten’s of it. It also mean that each time we deploy an airflow we will need to deploy a new cache server which should be 2 thing completely independent.
> * I know this cache server is currently needed for airflow but on a long term we could easily imagine having the need for it for other usercase.
>   Why not having a distributed cache service for everything? We could imagine multiple server/container with one leader to get info from armada and updating all the follower with a Raft protocol. Read could then be scale out by just adding more container.
>   it could also be any other distributed cache / ephemeral storage.

It seems to me that there is a lot of questions about the deployment of this server.  I believe Chris and Albin suggested deployment in airflow as that was expected to have smaller amount of users than the Armada directly.  I don't know if that is true though.  Since we are pushing airflow for GR we may be support a large amount of individual users for airflow 
> * Authorization: it’s good to think about security, however the cache server if I understand correctly will be read only and should not have confidential info, just job name and status. So we could imagine have a v1 one without security and v2 with security. The security could be implemented by adding to the cache a field with the owner of the job that can fetch job status

In general, I think security will be handled by Armada.  We will probably use whatever Authorization is present in Armada.  We are going to directly use the grpc client so that should use the same auth as Armada.  @d80tb7 or @severinson Does this sound correct?
> * Airflow also fetch log of the different task in real time (to be verified). How do you plan to address this with the cache server / job status service?

That wasn't in the initial request.  I think storing logs may be very difficult in this design.  We are going to use redis as our data-store and the hope is that the payload won't be very large.  I think putting unbounded log data into the cache server is a nonstarter.  

@d80tb7 and @severinson, what are your thoughts on this?  I know we use a special service called binoculars for logs for lookout.  I 
> * Where the cache server fit in the process for airflow to cancel / retry tasks? (worst use case will be to have airflow to cancel huge dag with hundred/thousands of tasks started)

To be clear, the cache server is primarily saving events.  This should be handled without issue.  We are primarly focused on subscribing to events and we are updating the cache whenever a new event comes through.  So if you cancelled a bunch of tasks, it should be sent to the cache server.  
> * Devil advocate: why going the route of a separate service? Could we not improve the current service by supporting also pulling and potentially implementing a distributed cache there directly?
> * It will be good to have few word on the reason why armada will not scale / support high pull load? What are the technical reasons that motivate us of doing this.

@d80tb7 or @severinson could you please help answer these two questions?
> * What should be the behaviour if the entry is not in the cache?

@d80tb7 and I were discussing this on Friday.  There are a few options.
1) We code the operator to be tolerant of this so it would wait until the cache is primed.  
2) We fallback to the event client and directly read the event.

Remember, the cache's job is to subscribe to events in the jobset.  So we just have to call the event and read the latest event for every job in the jobset.
> * How long will you keep entry in the cache? Should it be tunable?
> 
I think that needs some thought.  If nothing in the jobset has been updated, we could purge the cache.  I do think it would be good to make that as a config.  Good idea @michelsumbul.
> For info, we have an internal process to capture design decision call ADR at GR. Once we have this doc in good shape and some consensus, we should put it in the ADR and get formal approval from armada/architect team.
> 
> Michel

**Comment by d80tb7** [comment]
Let me answer this first as it's the most important point:

>    It will be good to have few word on the reason why armada will not scale / support high pull load? What are the technical reasons that motivate us of doing this.

Consider the following
- Armada can mange (at least) 1 million cores
- Armada should therefore support (at least) 1 Million concurrent jobs

The model that Airflow wants to use is that each job is independent and that everyone should poll to see if their job has finished.  In  the case of Armada that would mean supporting 1 million simultaneous TCP connections and handling maybe 100K requests a second (assuming a 10 second polling interval).  Such a system is possible to make, but it would certainly be a technical challenge.  Instead, the original Armada developers went a different direction, with their key insights being the following observations:
 * Most users don't submit a single job, but a group of related job (a jobset)
 * In any given poll period the job status is unlikely to have changed.
 
 Realizing this, they designed a system  a) around events, rather than polling and b) allowed event subscriptions only at the jobset level rather than the job level. To see the benefits of this, assume that each jobset contains 1000 jobs and that the average job takes 30 mins to complete.  Now our system only needs to support 1000 concurrent TCP connections and the 100K polls a second have been replaced with  500 messages/second.  This is a somewhat simpler system to design.

Unfortunately, however, this means that Armada and Airflow have different views of the world.  This component aims to bridge the gap in the simplest way possible.

Now that is answered, onto other considerations:

 > The job service / caching should be an optional deployment of armada and not part of airflow. If it’s part of the airflow it will be more difficult to get it accepted by the airflow community. The airflow operator should be a simple client, like any client for any other system. I don’t think we want to put any complexity in the client. What do you think?

There is no proposal to make this service part of Airflow or have it merged into the Airflow project.  It will be an optional Armada component.  We do however want it to be *deployed* alongside Airflow
The main point of this component is indeed to make the client as simple as possible.  It will allow a simple, polling model which is what Airflow expects.

 >   What do you mean concretely by one cache per user/aiflow? Is it one cache structure in memory per client in a single server or multiple cache server instance?
    Having one per airflow mean we could easily have ten’s of it. It also mean that each time we deploy an airflow we will need to deploy a new cache server which should be 2 thing completely independent.
 Yes the proposal is that you deploy one per Airflow instance and if you want to deploy 10s of airflow instances then you deploy a separate job cache alongside each one.  The reasons for this are:
 1. It simplifies the permissions model
 2. It means that the scaling occurs naturally as more Airflow instances are added.
 3. It reinforces the point that for now this is is simply a way to make airflow work.
     
    
>    I know this cache server is currently needed for airflow but on a long term we could easily imagine having the need for it for other usercase.
    Why not having a distributed cache service for everything? We could imagine multiple server/container with one leader to get info from armada and updating all the follower with a Raft protocol. Read could then be scale out by just adding more container.
    it could also be any other distributed cache / ephemeral storage.
    
This is really answered in my initial description of how Armada's design went in a different direction to that which Airflow expects.  If we find lots of other components/use cases that want a per job/polling model, then it may be that Armada would need to be redesigned to accommodate this.  For now, I would ague that proving Armada-Aiflow integration with a relatively simple component is a better idea that redesigning Armada itself.

>    Authorization: it’s good to think about security, however the cache server if I understand correctly will be read only and should not have confidential info, just job name and status. So we could imagine have a v1 one without security and v2 with security. The security could be implemented by adding to the cache a field with the owner of the job that can fetch job status

The cache server should be pretty light on security from a read point of view although it will contain error messages which may contain some sensitive information if the company using Armada cares about protecting Core IP. 

>    Airflow also fetch log of the different task in real time (to be verified). How do you plan to address this with the cache server / job status service?

This will need some thought.  Armada logs are K8s logs.  I don't know how Airflow wants to retrieve logs, but we probably want to avoid the situation where all Airflow jobs are polling K8s for logs as  that won't work at the scales Armada needs to support.

 >   Where the cache server fit in the process for airflow to cancel / retry tasks? (worst use case will be to have airflow to cancel huge dag with hundred/thousands of tasks started)

The airflow operator should be able to use existing armada apis to do this- the cache will play no role here.

 >   Devil advocate: why going the route of a separate service? Could we not improve the current service by supporting also pulling and potentially implementing a distributed cache there directly?
 
 Because changing Armada for retrieving job statuses per jobset to changing job statuses per job is a big change.  By breaking it into a separate service specially designed for airflow we make it clear that it is limited in scope to just that use case.

thanks,

Chris

**Review by jayofdoom — COMMENTED** [review]
There seems to have been consensus reached in this, it can likely merge now with any further changes to come via PR to the doc.

**Review by kannon92 — APPROVED** [review]
From GR, this was approved internally.

**Pull request merged**